fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.h

```diff
@@ -77,47 +77,40 @@ class NormalPriorPose3DCostFunctor
    *              (x, y, z, qx, qy, qz)
    * @param[in] b The 3D pose measurement or prior in order (x, y, z, qw, qx, qy, qz)
    */
-  NormalPriorPose3DCostFunctor(const fuse_core::Matrix6d& A, const fuse_core::Vector7d& b);
+  NormalPriorPose3DCostFunctor(const fuse_core::Matrix6d& A, const fuse_core::Vector7d& b):
+  A_(A),
+  b_(b),
+  orientation_functor_(fuse_core::Matrix3d::Identity(), b_.tail<4>())  // Delta will not be scaled
+  {
+  }
 
   /**
    * @brief Evaluate the cost function. Used by the Ceres optimization engine.
    */
   template <typename T>
-  bool operator()(const T* const position, const T* const orientation, T* residual) const;
+  bool operator()(const T* const position, const T* const orientation, T* residual) const {
+    // Compute the position error
+    residual[0] = position[0] - T(b_(0));
+    residual[1] = position[1] - T(b_(1));
+    residual[2] = position[2] - T(b_(2));
+
+    // Use the 3D orientation cost functor to compute the orientation delta
+    orientation_functor_(orientation, &residual[3]);
+
+    // Scale the residuals by the square root information matrix to account for
+    // the measurement uncertainty.
+    Eigen::Map<Eigen::Matrix<T, 6, 1>> residual_map(residual);
+    residual_map.applyOnTheLeft(A_.template cast<T>());
+
+    return true;
+  }
 
 private:
   fuse_core::Matrix6d A_;
   fuse_core::Vector7d b_;
 
   NormalPriorOrientation3DCostFunctor orientation_functor_;
 };
-
-NormalPriorPose3DCostFunctor::NormalPriorPose3DCostFunctor(const fuse_core::Matrix6d& A, const fuse_core::Vector7d& b) :
-  A_(A),
-  b_(b),
-  orientation_functor_(fuse_core::Matrix3d::Identity(), b_.tail<4>())  // Delta will not be scaled
-{
-}
-
-template <typename T>
```

Comment:
TBH I think this templated function/operator could be here - same for the other class.

The constructor isn't templated, so that one should probably be inside the class definition or in a cpp file. So for that one I actually understand it has to be moved to avoid compilation errors.

Either way, if this works, that's ok for me. :smiley: